cmd/containerboot/main.go

```diff
@@ -0,0 +1,426 @@
+
+//go:build linux
+// +build linux
+
+// The containerboot binary is a wrapper for starting tailscaled in a
+// container. It handles reading the desired mode of operation out of
+// environment variables, bringing up and authenticating Tailscale,
+// and any other kubernetes-specific side jobs.
+//
+// As with most container things, configuration is passed through
+// environment variables. All configuration is optional.
+//
+//  - TS_AUTH_KEY: the authkey to use for login.
+//  - TS_ROUTES: subnet routes to advertise.
+//  - TS_DEST_IP: proxy all incoming Tailscale traffic to the given
+//                destination.
+//  - TS_TAILSCALED_EXTRA_ARGS: extra arguments to 'tailscaled'.
+//  - TS_EXTRA_ARGS: extra arguments to 'tailscale up'.
+//  - TS_USERSPACE: run with userspace networking (the default)
+//                  instead of kernel networking.
+//  - TS_STATE_DIR: the directory in which to store tailscaled
+//                  state. The data should persist across container
+//                  restarts.
+//  - TS_ACCEPT_DNS: whether to use the tailnet's DNS configuration.
+//  - TS_KUBE_SECRET: the name of the Kubernetes secret in which to
+//                    store tailscaled state.
+//  - TS_SOCKS5_SERVER: the address on which to listen for SOCKS5
+//                      proxying into the tailnet.
+//  - TS_OUTBOUND_HTTP_PROXY_LISTEN: the address on which to listen
+//                                   for HTTP proxying into the tailnet.
+//  - TS_SOCKET: the path where the tailscaled local API socket should
+//               be created.
+//  - TS_AUTH_ONCE: if true, only attempt to log in if not already
+//                  logged in. If false (the default, for backwards
+//                  compatibility), forcibly log in every time the
+//                  container starts.
+//
+// When running on Kubernetes, TS_KUBE_SECRET takes precedence over
+// TS_STATE_DIR. Additionally, if TS_AUTH_KEY is not provided and the
+// TS_KUBE_SECRET contains an "authkey" field, that key is used.
+package main
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io/fs"
+	"log"
+	"net/netip"
+	"os"
+	"os/exec"
+	"os/signal"
+	"strconv"
+	"strings"
+	"time"
+
+	"golang.org/x/sys/unix"
+	"tailscale.com/client/tailscale"
+	"tailscale.com/ipn/ipnstate"
+)
+
+func main() {
+	log.SetPrefix("boot: ")
+	tailscale.I_Acknowledge_This_API_Is_Unstable = true
+
+	cfg := &settings{
+		AuthKey:         defaultEnv("TS_AUTH_KEY", ""),
+		Routes:          defaultEnv("TS_ROUTES", ""),
+		ProxyTo:         defaultEnv("TS_DEST_IP", ""),
+		DaemonExtraArgs: defaultEnv("TS_TAILSCALED_EXTRA_ARGS", ""),
+		ExtraArgs:       defaultEnv("TS_EXTRA_ARGS", ""),
+		InKubernetes:    os.Getenv("KUBERNETES_SERVICE_HOST") != "",
+		UserspaceMode:   defaultBool("TS_USERSPACE", true),
+		StateDir:        defaultEnv("TS_STATE_DIR", ""),
+		AcceptDNS:       defaultBool("TS_ACCEPT_DNS", false),
+		KubeSecret:      defaultEnv("TS_KUBE_SECRET", "tailscale"),
```

Comment:
you can't default this here, as is you will fail on line 90

Reply:
when running outside k8s

Reply:
Gaaah I want my tristate values :(

Reply:
Fixed, l90 now checks the envvar directly, so that initialization block still captures all the defaults for everything.